### PR TITLE
Set channel permissions for owner during m.carrier_add

### LINKED ISSUE #93: Add permissions setup to channel created by m.carrier_add
Requires inputting Discord name/ID of carrier's owner

### PR DESCRIPTION
**Pull request by cimspin**
Requires Discord.py version 1.7 or higher
Fixes #93

### CODE REVIEW

File: MissionAlertBot.py [hunk 1]
```diff
@@ -1293,6 +1295,30 @@ async def carrier_add(ctx, short_name, long_name, carrier_id, owner_id):
 
     if not channel:
         raise EnvironmentError(f'Could not create carrier channel {stripped_name.lower()}')
+        return
```
Comment (gcruicks) [inline MissionAlertBot.py:1298]: No need for return here, raise exits this method directly

File: MissionAlertBot.py [hunk 2]
```diff
@@ -1293,6 +1295,30 @@ async def carrier_add(ctx, short_name, long_name, carrier_id, owner_id):
...
+                                            attach_files=True,
+                                            add_reactions=True,
+                                            external_emojis=True,
+                                            manage_messages=True,
+                                            read_message_history=True,
+                                            use_slash_commands=True)
+        print(f"Set permissions for {owner} in {channel}")
+    except:
+        raise EnvironmentError(f'Could not set channel permissions for {owner.display_name} in {channel}')
+        return
```
Comment (gcruicks) [inline MissionAlertBot.py:1321]: no need for return

File: MissionAlertBot.py [hunk 3]
```diff
@@ -1293,6 +1295,30 @@ async def carrier_add(ctx, short_name, long_name, carrier_id, owner_id):
 
     if not channel:
         raise EnvironmentError(f'Could not create carrier channel {stripped_name.lower()}')
+        return
+
+    # add carrier owner to channel permissions
+
+    owner = await bot.fetch_user(owner_id)
```
Comment (gcruicks) [inline MissionAlertBot.py:1302]: What gets returned if the owner is not found? I guess the exception block below will capture that.

File: MissionAlertBot.py [hunk 4]
```diff
@@ -1293,6 +1295,30 @@ async def carrier_add(ctx, short_name, long_name, carrier_id, owner_id):
...
+                                            send_messages=True,
+                                            embed_links=True,
+                                            attach_files=True,
+                                            add_reactions=True,
+                                            external_emojis=True,
+                                            manage_messages=True,
+                                            read_message_history=True,
+                                            use_slash_commands=True)
+        print(f"Set permissions for {owner} in {channel}")
+    except:
```
Comment (gcruicks) [inline MissionAlertBot.py:1319]: Do we know what exception gets raised? Better to catch what we expect, rather than everything

**Review by gcruicks — CHANGES_REQUESTED** [review]
Remove the returns statements that are redundant. Maybe catch the specific exception expected.

File: MissionAlertBot.py (hunk 3 above)
Comment (cimspin) [reply MissionAlertBot.py:1302]: Added try/except for bot.fetch_user

File: MissionAlertBot.py (hunk 4 above)
Comment (cimspin) [reply MissionAlertBot.py:1319]: I don't know what would get raised (or how to catch a specific exception). Problems I had when testing were (1) discord.py version too low for the use_slash_commands permission and (2) bot not having sufficient permissions (unfortunately it needs full on admin permissions to give another user (i.e. the CO) the fully elevated channel permissions that we grant to COs in their channels)

Comment (gcruicks) [reply MissionAlertBot.py:1324]: https://discordpy.readthedocs.io/en/stable/api.html#discord.abc.GuildChannel.set_permissions

Raises

- Forbidden – You do not have permissions to edit channel specific permissions.
- HTTPException – Editing channel specific permissions failed.
- NotFound – The role or member being edited is not part of the guild.
- InvalidArgument – The overwrite parameter invalid or the target type was not Role or Member.

I would start by catching these exceptions only

Comment (cimspin) [reply MissionAlertBot.py:1324]: I don't know how to do this. Is it like `except Forbidden:` for each error type? How do I catch multiple error types? Just keep putting new `except:` instances after the first?

Comment (gcruicks) [reply MissionAlertBot.py:1324]: except (exeption1, exception2, exception3):

**Review by gcruicks — APPROVED** [review]
Ship it!

**Pull request merged**